packages/playwright-core/src/server/frames.ts

```diff
@@ -1318,6 +1318,33 @@ export class Frame extends SdkObject {
     }, this._page._timeoutSettings.timeout({}));
   }
 
+  async viewportRatio(metadata: CallMetadata, selector: string, options: types.StrictOptions = {}): Promise<number> {
+    const controller = new ProgressController(metadata, this);
+    return controller.run(async progress => {
+      progress.log(`  getting viewport ratio of ${this._asLocator(selector)}`);
+      const pair = await this.resolveFrameForSelectorNoWait(selector, options);
+      if (!pair)
+        return 0;
```

Comment:
You need to `retryWithProgressAndTimeouts` here, in case locator does not resolve immediately. Take a look at `_callOnElementOnceMatches` for example.